### PR TITLE
Enhancement: Allow Widgets in Toast Title and Description

### PR DESCRIPTION
**Pull request by callmephil**
Description:
This pull request addresses issue #61 and introduces the ability to include widgets within the title or description when displaying a toast using Toastification.

Notes:
This is a work in progress! Most of the core changes are done and the app is runnable. Feel free to review =) 

# Work done
- [x] Replace Title and Description with Widget.
- [X] Update Example 
- [x] Update Readme 

# In Progress
- [x] Re-implement default Text styling.
- [x] Provide an example with RichText

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,62 +1,92 @@
+## 1.1.1
```
Comment (payam-zahedi) [inline CHANGELOG.md:1]: Please do not change this file. I will update it whenever we have new release.

File: example/lib/src/features/home/views/ui_states/toast_code_formatter.dart [hunk 2]
```diff
@@ -18,7 +18,7 @@ class ToastCodeFormatter {
     title: '${toastDetail.title}',
   ''');
 
-    if (toastDetail.description.isNotEmpty) {
+    if (toastDetail.description != null) {
       code.writeln('\tdescription: \'${toastDetail.description}\',');
```
Comment (payam-zahedi) [inline example/lib/src/features/home/views/ui_states/toast_code_formatter.dart:22]: Are you sure is this working? 

I think you should change it to this
``` dart
code.writeln('\tdescription: Text(\'${toastDetail.description}\'),');
```

Comment (payam-zahedi) [reply example/lib/src/features/home/views/ui_states/toast_code_formatter.dart:22]: Also for the title

File: example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart [hunk 3]
```diff
@@ -32,8 +32,8 @@ class ToastDetail with _$ToastDetail {
     @Default(ToastificationStyle.flat) ToastificationStyle style,
     // placement
     @Default(Alignment.topLeft) AlignmentGeometry alignment,
-    @Default('Component updates available.') String title,
-    @Default('Component updates available.') String description,
+    @Default(Text('Component updates available.')) Widget? title,
+    @Default(Text('Component updates available.')) Widget? description,
```
Comment (payam-zahedi) [inline example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart:36]: Why do we need to make them nullable?

File: lib/src/core/toastification.dart [hunk 4]
```diff
@@ -218,9 +218,9 @@ class Toastification {
     ToastificationAnimationBuilder? animationBuilder,
     ToastificationType? type,
     ToastificationStyle? style,
-    required String title,
+    Widget? title,
```
Comment (payam-zahedi) [inline lib/src/core/toastification.dart:221]: I'm not sure about making title nullable for now. 

Do you have any reason for that?

**Review by payam-zahedi — COMMENTED** [review]
Thank you so much @callmephil 

I wrote my comments on your changes 

If you need anything just letme know

File: lib/src/core/toastification.dart (hunk 4 above)
Comment (callmephil) [reply lib/src/core/toastification.dart:221]: Maybe someone doesn't want to have a title for example have extra space for the description.

File: example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart (hunk 3 above)
Comment (callmephil) [reply example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart:36]: To match the intended model struct of the package

File: example/lib/src/features/home/views/ui_states/toast_code_formatter.dart (hunk 2 above)
Comment (callmephil) [reply example/lib/src/features/home/views/ui_states/toast_code_formatter.dart:22]: I will try this out thanks!

File: CHANGELOG.md (hunk 1 above)
Comment (callmephil) [reply CHANGELOG.md:1]: Sure I'll remove it, do you believe the changelog is good enough?

**Comment by callmephil** [comment]
> Thank you so much @callmephil
> 
> I wrote my comments on your changes
> 
> If you need anything just letme know

That's awesome thank you for reviewing! For the styling I'm considering moving it directly to the theme in the example as per best practice. It will probably ennoy some people is that Ok?
Otherwise I'll need an extra flag "keepOldTextStyling" or something like that.

File: example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart (hunk 3 above)
Comment (callmephil) [reply example/lib/src/features/home/views/ui_states/toast_detail_ui_state.dart:36]: To clarify Nullable is commonly used by material and Cupertino devs when it comes to building U.I components it provides greater Flexibility to the dev. 

In this case, this allows us to provide more vertical space for the description field, or/and gain a little performance by not having to initialize an empty text Widget.

File: lib/src/core/toastification.dart (hunk 4 above)
Comment (callmephil) [reply lib/src/core/toastification.dart:221]: To clarify Nullable is commonly used by material and Cupertino devs when it comes to building U.I components it provides greater Flexibility to the dev.

In this case, this allows us to provide more vertical space for the description field, or/and gain a little performance by not having to initialize an empty text Widget.

Comment (payam-zahedi) [reply lib/src/core/toastification.dart:221]: Yes, you are right

**Comment by payam-zahedi** [comment]
> > Thank you so much @callmephil
> > I wrote my comments on your changes
> > If you need anything just letme know
> 
> That's awesome thank you for reviewing! For the styling I'm considering moving it directly to the theme in the example as per best practice. It will probably ennoy some people is that Ok? Otherwise I'll need an extra flag "keepOldTextStyling" or something like that.

What do you mean by `styling`?

**Comment by callmephil** [comment]
> > > Thank you so much @callmephil
> > > I wrote my comments on your changes
> > > If you need anything just letme know
> > 
> > 
> > That's awesome thank you for reviewing! For the styling I'm considering moving it directly to the theme in the example as per best practice. It will probably ennoy some people is that Ok? Otherwise I'll need an extra flag "keepOldTextStyling" or something like that.
> 
> What do you mean by `styling`?

to answer your question since the description and title were string by default the styling of text was hardcoded it's no longer the case and for the demo, the styling for the text needs to be applied on the 'example' level instead.

Sorry for the late reply, I'm quite busy at work. I'll finish the work asap.

**Comment by payam-zahedi** [comment]
Alright

So you can add two other fields to the `show` method for styles like `titleStyle` and `descriptionStyle`.
and you can add them in the README.md. (I can do it also if you don't have enought time for it)

but, please don't add anything to the [example project ](https://github.com/payam-zahedi/toastification/tree/main/example)([ToastBuilder](https://payamzahedi.com/toastification/)). I would prefer to talk to [Sepideh](https://sepide.design/)(toastification designer) before any changes in the UI of the ToastBuilder app.

**Comment by callmephil** [comment]
> Alright
> 
> So you can add two other fields to the `show` method for styles like `titleStyle` and `descriptionStyle`. and you can add them in the README.md. (I can do it also if you don't have enough time for it)
> 
> but, please don't add anything to the [example project ](https://github.com/payam-zahedi/toastification/tree/main/example)([ToastBuilder](https://payamzahedi.com/toastification/)). I would prefer to talk to [Sepideh](https://sepide.design/)(toastification designer) before any changes in the UI of the ToastBuilder app.

Hey! 

for the Text style title and description, It won't make much sense, since these fields are Widgets now so applying text style won't fit all scenarios. 
Regarding the change in the example, It won't be anything different that what it is now.  It's just that the current text style is hardcoded in the library and it needs to be moved to the example. 

I had hectic weeks recently I'll work on that soon. Maybe it will be less confusing when you see the code.

File: lib/src/widget/built_in/built_in.dart [hunk 5]
```diff
@@ -84,44 +83,32 @@ class BuiltInContent extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
-    Widget content = Text(
-      title,
-      style: style.titleTextStyle(context)?.copyWith(
-            color: foregroundColor,
-          ),
-    );
+    Widget? content = title ?? const SizedBox();
```
Comment (Maksimka101) [inline lib/src/widget/built_in/built_in.dart:86]: Hello, @callmephil 

Looks like you forgot to wrap the title widget with the `DefaultTextStyle`. At the moment, title text has this style: `style.titleTextStyle(context)?.copyWith(color: foregroundColor)` 🙂

Comment (payam-zahedi) [reply lib/src/widget/built_in/built_in.dart]: Thanks @Maksimka101 for your review 

I didn't recognize this one 

Fixed

**Review by payam-zahedi — APPROVED** [review]
@callmephil Thanks for your contribution

I updated the code and now its ready to merge

**Pull request merged**